added changeRoutine, enterRoutine and leaveRoutine

ArnoldSmith86 opened the pull request:
This was removed from #197 when I realized that every writing access to widget states now has to be async.

This adds new properties to all widgets. They are each an array of operations just like `clickRoutine`:

1. `changeRoutine` which gets called whenever a property of the widget it is in changes. The routine will have preset variables `property`, `oldValue` and `value`.
2. `enterRoutine` which gets called whenever a new child gets added to the widget it is in. The routine will have the preset variable `oldParentID` and the preset collection `child`.
3. `leaveRoutine` which gets called whenever a child gets removed from the widget it is in. The routine will have the preset collection `child`.

In addition you can define changeRoutines for a single property. Just add "ChangeRoutine" to the property you want to monitor. The routines will have preset variables `oldValue` and `value`.<br>For example: `xChangeRoutine`, `parentChangeRoutine`, `activeFaceChangeRoutine`, `MY_CUSTOM_PROPERTYChangeRoutine`

To do:

- [x] Change routine defaults in the JSON editor from `null` to `[]`.
- [x] Fix `applyVariables` in JSON editor.
- [x] Fix bug with updating in JSON editor.

ArnoldSmith86 commented:
PR-SERVER-BOT: You can play around with it here: http://212.47.248.129:3406/pr-test (or any other room on that server)

rogerl50 commented:
Is the idea to put `"enterRoutine"` and `"leaveRoutine"` in a holder (say), and that they should be triggered when a widget enters/leaves the holder? I tried that, and as far as I can tell, the button click that should have been triggered by those routines never happens. Look at http://212.47.248.129:3406/poker and the hand (the holder at the bottom of the screen). I assume I'm just misinterpreting how this is to be used.

96LawDawg commented:
> Is the idea to put `"enterRoutine"` and `"leaveRoutine"` in a holder (say), and that they should be triggered when a widget enters/leaves the holder? I tried that, and as far as I can tell, the button click that should have been triggered by those routines never happens. Look at http://212.47.248.129:3406/poker and the hand (the holder at the bottom of the screen). I assume I'm just misinterpreting how this is to be used.

It works.  You used GET instead of SELECT for the button id.  It is off by 1 card, but that is probably the order it is being triggered.

rogerl50 commented:
Look at http://212.47.248.129:3406/poker. Click on the Sit button in one of the unoccupied seats, then drag a card into the holder. The count remains at zero. Drag in a second card, and the count changes to 1. (The count is the little white circle near your cards). Drag that card back out, and the count remains at 1. Drag out the second card and it changes back to zero.

And, BTW, if you drag that second card out and back in without dropping it, the count remains at zero.

ArnoldSmith86 commented:
Yeah, I'll have a look tomorrow. You could also test `changeRoutine` on the cards.

ArnoldSmith86 commented on client/components/deckEditor.js:
```diff
@@ -34,7 +34,7 @@ export const deckEditor = {
       },
 
-      _addCardCallback(imagePath, fileName) {
+      async addCardCallback(imagePath, fileName) {
```
Oops.

96LawDawg commented:
> Look at http://212.47.248.129:3406/poker. Click on the Sit button in one of the unoccupied seats, then drag a card into the holder. The count remains at zero. Drag in a second card, and the count changes to 1. (The count is the little white circle near your cards). Drag that card back out, and the count remains at 1. Drag out the second card and it changes back to zero.
> 
> And, BTW, if you drag that second card out and back in without dropping it, the count remains at zero.

I think the count problem has to do with state change updates when dragging cards.  If you use a move function to do the same thing, the count works.  But dragging the card, the count lags by one.  I think this is because the onLeave routine triggers before the card has "left" the holder.  If I remember right, @ArnoldSmith86 said that a dragged card does not lose it's parent until you let go of it.  Might have to revisit that for this scenario.

rogerl50 commented:
> > Look at http://212.47.248.129:3406/poker. Click on the Sit button in one of the unoccupied seats, then drag a card into the holder. The count remains at zero. Drag in a second card, and the count changes to 1. (The count is the little white circle near your cards). Drag that card back out, and the count remains at 1. Drag out the second card and it changes back to zero.
> > And, BTW, if you drag that second card out and back in without dropping it, the count remains at zero.
> 
> I think the count problem has to do with state change updates when dragging cards. If you use a move function to do the same thing, the count works. But dragging the card, the count lags by one. I think this is because the onLeave routine triggers before the card has "left" the holder. If I remember right, @ArnoldSmith86 said that a dragged card does not lose it's parent until you let go of it. Might have to revisit that for this scenario.

There are now two buttons in that room. The left-hand one moves a card from the upper left holder to the hand, and the right-hand one does the reverse. I'm observing identical behavior to dragging the cards.

96LawDawg commented:
> There are now two buttons in that room. The left-hand one moves a card from the upper left holder to the hand, and the right-hand one does the reverse. I'm observing identical behavior to dragging the cards.

For me, the button moving to the hand is still off by one, but the button moving from the hand shows an accurate count.  I think this is still consistent with my hypothesis about this being related which holder is the parent of the object at the exact time that the enter or leave routine is triggered.  It would be the original holder when moving to the hand (even when doing it with a button) which is why it doesn't update that way.

96LawDawg commented:
On the other hand, maybe it is something with the poker room.  I just started doing my own tests.  See [http://212.47.248.129:3406/Go](http://212.47.248.129:3406/Go).  The holder has all the commands in the enterRoutine.  The black holder clicks on the sum button (the way you used to have to count pieces).  Both work the same and both work as expected by incrementing the count as soon as the piece is dropped in the holder.

96LawDawg commented:
Further testing is proving very successful.  I think I found a solution for your request in Quad Line 3D to lock down the pieces.  Check out the upper row in http://212.47.248.129:3406/QuadLine3D.  Also my huge routine in Crafty (4000 lines with multiple compute, select, get, set, and random functions) worked just fine.  More testing to do, but I wanted to show the Quad Line thing as an example of how powerful this is.

96LawDawg commented:
In the main pr-test room, I have a demo of changeRoutine.  I guess I wasn't expecting it to be as "sensitive" as it is.  If **anything** changes, then the routine is triggered.  So it probably would not be useful to say "if anything changes, do x."  So you need IF statements to limit what the changeRoutine does.

Another way around this problem might be to have a changeRoutine parameter.  The parameter (or multiple parameters) tell changeRoutine what to look for.  Only if one of those things changes (like parent, or face, or rotation) would the routine kick it.  Because right now if you breathe too hard on the computer screen, the changeRoutine activates (but that is solvable through IF statements).

I only found one bug so far.  When trying to remove anything in a room on this test server (even a room without the new features), the room crashes if you remove it through the JSON editor.  Can still remove the old way through the old editor.

rogerl50 commented:
I finally have a small test case that exhibits a problem with `enterRoutine`. Take a look at http://212.47.248.129:3406/we3v. If you drag a card from the deck into the (empty) holder, the lower counter will read 0. If you click on the `DEAL` button, the upper counter will read 1. I believe that the `clickRoutine` of the button and the `enterRoutine` of the holder should do the same thing; namely, count the cards belonging to me that are in the holder.

ArnoldSmith86 commented:
> As a side note, related or not, the `enterRoutine` seems to start life with a `child` collection, which I did not define, that contains the offending card.

That's a feature.

> Take a look at http://212.47.248.129:3406/we3v. If you drag a card from the deck into the (empty) holder, the lower counter will read 0.

This should be fixed now. Please continue testing and report any problems.

rogerl50 commented:
> That's a feature.

What exactly is the child collection? It always seems to contain just the first card dragged in (in my example).

> This should be fixed now. Please continue testing and report any problems.

This did fix the issue in my room.

ArnoldSmith86 commented:
> What exactly is the child collection? It always seems to contain just the first card dragged in (in my example).

It is the widget that got added to the holder. If multiple widgets get added, the routine will be called for each one.

96LawDawg commented:
I did find one other strange thing.  In http://212.47.248.129:3406/pr-test, the blue piece will report multiple changes happening at the same time. Drag it around and you can see x and y changing. But if you click on it, the face updates properly, but the rotation does not. But then when you move it left or right or up or down, then the rotation property gets triggered to get updated.  However, when you click the button to update both at the same time, they work.  So clicking directly on the object gave a different result than clicking on a button to do the same thing.

96LawDawg commented:
There are definitely some strange things while using the JSON editor.  In http://212.47.248.129:3406/Label, sometimes I try to type in the label box (which has a changeRoutine in it) while the JSON editor is open.  After the first letter, the label loses focus and the typing happens in the editor.  I kick myself, then delete the newly added text in the editor.  But then the changeRoutine for that label doesn't work any more.  Even if I refresh the room, it still doesn't work for me, but it did work for someone else.  The other labels still work fine.

ArnoldSmith86 commented:
> I did find one other strange thing. In http://212.47.248.129:3406/pr-test, the blue piece will report multiple changes happening at the same time. Drag it around and you can see x and y changing. But if you click on it, the face updates properly, but the rotation does not. But then when you move it left or right or up or down, then the rotation property gets triggered to get updated. However, when you click the button to update both at the same time, they work. So clicking directly on the object gave a different result than clicking on a button to do the same thing.

I agree that it is strange behaviour. It's basically a bug within a bug that nearly cancel each other out. Changing the `activeFace` of a `BasicWidget` should not trigger change events for all the properties defined inside `faces`. The properties shouldn't get copied to the widget either which currently happens at some point. For example: if each face has an `image` defined, the widget itself shouldn't need to define one. Whenever the `image` property gets queried it should just look into the currently active face and return that one.

ArnoldSmith86 commented:
> After the first letter, the label loses focus and the typing happens in the editor.

This is actually not related to this PR but I only realized that after fixing it. So it's part of this PR now.

> But then the changeRoutine for that label doesn't work any more.

Let me see if I can reproduce that.

RaphaelAlvez commented:
leaveRoutine is being executed before removing ownership of a card

ArnoldSmith86 commented:
Do we want to add change routines per property? For example `parentChangeRoutine`?

RaphaelAlvez commented:
Awesome idea. With that should also go a variable with the new and old values. Ideally this would be built in a way that it only reads the change routines name and chooses a property to take care. So even custom properties could be used

96LawDawg commented:
Good idea. That would address my comment above about the routine being sensitive to everything. If it “listens” only for a single parameter then that would help some cases.

ArnoldSmith86 commented:
> With that should also go a variable with the new and old values.

That's already the case.

> Ideally this would be built in a way that it only reads the change routines name and chooses a property to take care. So even custom properties could be used

Exactly.

ArnoldSmith86 commented:
I addressed every issue I'm aware of and updated the description to actually describe what the PR adds.

Please re-test the problems you were having and report anything that is broken.

96LawDawg commented:
Help!  Something in the recent commits made it impossible to use "applyVariables" in the JSON editor.  I can still manually type it in using the old editor, but I threw away my tools for carving in stone after the JSON editor came out.  I'm still working on the chat room files and can't make any more progress.

But on a good note, everything else I've tested so far (in particular parameter specific changeRoutines, labels losing focus, and JSON editor deletion problems) are all fine.  Nothing about enterRoutine or leaveRoutine seems out of place.  I think we are getting close on this.

96LawDawg commented:
More JSON errors to report.  Demo room: http://212.47.248.129:3406/JSONError.  Start the chat, pick any color.  Press options, then press Half button. Then press options multiple times. It flips back and forth between full and half size (the buttons are out of place; I'm working on that).  But when I open the JSON editor and click on a button (to watch variables change live), then try clicking options again, it freezes up the room.  See video.

https://user-images.githubusercontent.com/76912527/116642082-b0d35c00-a933-11eb-84d9-3519460462d4.mp4

ArnoldSmith86 commented:
I'll fix that. Just didn't have time this week.

ArnoldSmith86 commented:
> I addressed every issue I'm aware of[...].
> 
> Please re-test the problems you were having and report anything that is broken.

rogerl50 reviewed (approved):
Approved through test with my test cases.

96LawDawg reviewed (approved):
Approved through extensive testing.  Everything I have found causes a problem has been corrected. Ready for merge in my opinion.

bjalder26 commented:
I think leaveRoutine may be double firing.  In the following room if you put something in the blue holder, then remove it, then the widget with a colon on it will get cloned twice instead of once.  The leaveRoutine on the blue holder is set to clone the widget with the colon on it once. 

https://virtualtabletop.io/rgp2

rogerl50 commented:
> I think leaveRoutine may be double firing. In the following room if you put something in the blue holder, then remove it, then the widget with a colon on it will get cloned twice instead of once. The leaveRoutine on the blue holder is set to clone the widget with the colon on it once.
> 
> https://virtualtabletop.io/rgp2

You might want to post this in the discord #help channel, as this PR is closed. But in any case, I tried it, and when I moved the C in and out, the colon was cloned once, just as you'd expect. Did you mean something else?

bjalder26 commented:
> > I think leaveRoutine may be double firing. In the following room if you put something in the blue holder, then remove it, then the widget with a colon on it will get cloned twice instead of once. The leaveRoutine on the blue holder is set to clone the widget with the colon on it once.
> > https://virtualtabletop.io/rgp2
> 
> You might want to post this in the discord #help channel, as this PR is closed. But in any case, I tried it, and when I moved the C in and out, the colon was cloned once, just as you'd expect. Did you mean something else?

It was cloned twice, they just get the same offset so that you won't see the second one unless you move them.

rogerl50 commented:
You should open an issue. I added a button to your room with a clickRoutine the same as your leaveRoutine, and the clone only happens once. Also, the leaveRoutine fires if you just move the item inside the holder and don't actually take it out.
